master/static/srv/ship_logs.py

```diff
@@ -0,0 +1,355 @@
+
+    threads_started = []  # type: List[threading.Thread]
+    p = subprocess.Popen(
+        cmd, stdin=subprocess.DEVNULL, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+    )
+    # Just for mypy.
+    assert p.stdout and p.stderr
+
+    try:
+        # Set up signal forwarding.
+        def signal_passthru(signum: Any, frame: Any):
+            p.send_signal(signum)
+
+        for sig in [
+            signal.SIGINT,
+            signal.SIGTERM,
+            signal.SIGHUP,
+            signal.SIGUSR1,
+            signal.SIGUSR2,
+            signal.SIGWINCH,
+        ]:
+            signal.signal(sig, signal_passthru)
+
+        stdout = Collector(p.stdout, "stdout", emit_stdout_logs, metadata, logq, doneq)
+        stderr = Collector(p.stderr, "stderr", emit_stdout_logs, metadata, logq, doneq)
+        shipper = Shipper(master_url, token, cert_name, cert_file, logq, doneq)
+
+        stdout.start()
+        threads_started.append(stdout)
+        stderr.start()
+        threads_started.append(stderr)
+        shipper.start()
+        threads_started.append(shipper)
+
+        # expect 3 messages on the doneq
+        first_who = None
+        first_error = None
+        for _ in range(3):
+            who, error = doneq.get()
+            if who == "shipper":
+                # There's no point in collecting logs after the shipper is gone.
+                stdout.cancel()
+                stderr.cancel()
+            if first_error is None and error is not None:
+                first_who = who
+                first_error = error
+                # If our logging infrastructure ever crashes, just give up on the child process.
+                p.kill()
+
+        if first_error is not None:
+            raise RuntimeError(f"failure in log shipper; {first_who} thread died") from first_error
+
+        return p.wait()
+
+    finally:
+        p.kill()
+        p.wait()
+        for thread in threads_started:
+            thread.join()
+
+
+if __name__ == "__main__":
+    try:
+        master_url = os.environ["DET_MASTER"]
+        cert_name = os.environ.get("DET_MASTER_CERT_NAME", "")
+        cert_file = os.environ.get("DET_MASTER_CERT_FILE", "")
+        token = os.environ["DET_SESSION_TOKEN"]  # XXX: which env var to use?
+        raw_metadata = os.environ["DET_TASK_LOGGING_METADATA"]
+        try:
+            metadata = json.loads(raw_metadata)
+            assert isinstance(metadata, dict)
+        except Exception:
+            raise ValueError(f"invalid DET_TASK_LOGGING_METADATA: '{raw_metadata}'") from None
+
+        metadata["source"] = "task"
+
+        # XXX: metadata["container_id"] = os.environ.get("DET_CONTAINER_ID", "")?
+        # XXX: metadata["agent_id"] = os.environ["DET_AGENT_ID"]?
+
+        # XXX why is this not done master-side?
+        metadata.pop("trial_id", None)
+
+        emit_stdout_logs = bool(
+            distutils.util.strtobool(os.environ.get("DET_SHIPPER_EMIT_STDOUT_LOGS", "false"))
```

Comment:
nit: distutils is [going away](https://peps.python.org/pep-0632/#:~:text=In%20Python%203.10%20and%203.11,platforms%20will%20not%20be%20added.). may be nice to save us this bug in however long, since it will just break logging.

Reply:
removed.  This is the motivation I've been looking for to stop using that silly function and use "nonempty environment variable means true".